File: python/problems/leetcode/0743_network_delay_time/solution.py
"""LeetCode 743: Network Delay Time."""

from __future__ import annotations

import heapq


class Solution:
    def networkDelayTime(self, times: list[list[int]], n: int, k: int) -> int:
        adj: list[list[tuple[int, int]]] = [[] for _ in range(n)]
        for u, v, w in times:
            adj[u - 1].append((v - 1, w))
        inf = 10**9
        dist = [inf] * n
        dist[k - 1] = 0
        pq: list[tuple[int, int]] = [(0, k - 1)]
        while pq:
            d, u = heapq.heappop(pq)
            if d != dist[u]:
                continue
            for v, w in adj[u]:
                nd = d + w
                if nd < dist[v]:
                    dist[v] = nd
                    heapq.heappush(pq, (nd, v))
        mx = max(dist)
        return mx if mx < inf else -1


if __name__ == "__main__":
    assert Solution().networkDelayTime([[2, 1, 1], [1, 3, 1], [2, 3, 1]], 3, 2) == 1
    assert Solution().networkDelayTime([[2, 1, 1], [1, 3, 1], [2, 3, 1]], 4, 2) == -1
    assert Solution().networkDelayTime([[1, 2, 1]], 2, 1) == 1
    assert Solution().networkDelayTime([[1, 2, 1]], 2, 2) == -1
    print("0743 OK")

File: cpp/problems/leetcode/0743_network_delay_time/solution.cpp
// LeetCode 743: Network Delay Time
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<pair<int, int>>> adj(n);
        for (auto& e : times) adj[e[0] - 1].push_back({e[1] - 1, e[2]});
        const int inf = 1e9;
        vector<int> dist(n, inf);
        dist[k - 1] = 0;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
        pq.push({0, k - 1});
        while (!pq.empty()) {
            auto [d, u] = pq.top();
            pq.pop();
            if (d != dist[u]) continue;
            for (auto [v, w] : adj[u]) {
                int nd = d + w;
                if (nd < dist[v]) {
                    dist[v] = nd;
                    pq.push({nd, v});
                }
            }
        }
        int mx = *max_element(dist.begin(), dist.end());
        return mx < inf ? mx : -1;
    }
};

int main() {
    vector<vector<int>> t1{{2, 1, 1}, {1, 3, 1}, {2, 3, 1}};
    assert(Solution().networkDelayTime(t1, 3, 2) == 1);
    assert(Solution().networkDelayTime(t1, 4, 2) == -1);
    vector<vector<int>> t2{{1, 2, 1}};
    assert(Solution().networkDelayTime(t2, 2, 1) == 1);
    assert(Solution().networkDelayTime(t2, 2, 2) == -1);
    cout << "0743 OK" << endl;
    return 0;
}
